README.txt
                    Bitnami Magento Stack 1.9.1.0-1
                  ==============================


1. OVERVIEW

The Bitnami Project was created to help spread the adoption of freely
available, high quality, open source web applications. Bitnami aims to make
it easier than ever to discover, download and install open source software
such as document and content management systems, wikis and blogging
software.

You can learn more about Bitnami at https://bitnami.com

Magento is a web content management system (software for building and
maintaining a dynamic website) that keeps you behind the wheel. Magento's 
powerful feature-set means there's always a way to accomplish your vision.
Not only does it have all the features you'd expect: for instance galleries,
news, file downloads and community forums/chats, but it does so whilst
meeting the highest accessibility and professional standards. It is also
smart enough to go beyond page management, to automatically handle search
engine optimisation, and provide aggressive hack attack prevention. With
Magento, you get much more than just a Content Management System.

You can learn more about Magento at http://magento.com

The Bitnami Magento Stack is an installer that greatly simplifies the
installation of Magento and runtime dependencies. It includes ready-to-run
versions of Apache, MySQL and PHP. Magento Stack is distributed for free
under the Apache 2.0 license. Please see the appendix for the specific
licenses of all open source components included.

You can learn more about Bitnami Stacks at https://bitnami.com/stacks/

2. FEATURES

- Easy to Install

Bitnami Stacks are built with one goal in mind: to make it as easy as
possible to install open source software. Our installers completely automate
the process of installing and configuring all of the software included in
each Stack, so you can have everything up and running in just a few clicks.

- Independent

Bitnami Stacks are completely self-contained, and therefore do not interfere
with any software already installed on your system. For example, you can
upgrade your system's MySQL or Apache without fear of 'breaking' your
Bitnami Stack.

- Integrated

By the time you click the 'finish' button on the installer, the whole stack
will be integrated, configured and ready to go. 

- Relocatable

Bitnami Stacks can be installed in any directory. This allows you to have
multiple instances of the same stack, without them interfering with each other. 

3. COMPONENTS

Bitnami Magento Stack ships with the following software versions:

  - Magento 1.9.1.0
  - Apache 2.4.12
  - MySQL 5.5.42
  - PHP 5.4.37
  - phpMyAdmin 4.3.8

4. REQUIREMENTS

To install Bitnami Magento Stack you will need:

    - Intel x86 or compatible processor
    - Minimum of 256 MB RAM 
    - Minimum of 150 MB hard drive space
    - TCP/IP protocol support
    - Compatible operantig systems:
      - An x86 Linux operating system.
      - A 32-bit Windows operating system such as Windows 2000, XP, Vista
      or Windows Server 2003.
      - An OS X operating system x86.

5. INSTALLATION

The Bitnami Magento Stack is distributed as a binary executable installer.
It can be downloaded from:

https://bitnami.com/stacks/

The downloaded file will be named something similar to:

bitnami-magento-1.9.1.0-1-linux-installer.run on Linux or
bitnami-magento-1.9.1.0-1-linux-x64-installer.run on Linux 64 bit or
bitnami-magento-1.9.1.0-1-windows-installer.exe on Windows or
bitnami-magento-1.9.1.0-1-osx-x86-installer.dmg on OS X x86.

On Linux, you will need to give it executable permissions:

chmod 755 bitnami-magento-1.9.1.0-1-linux.run


To begin the installation process, double-click on that file, and you will
be greeted by the 'Welcome' screen. Pressing 'Next' will take you to the
Component Selection screen, where you can select the phpMyAdmin component. This
tool is intended to handle the administration of MySQL over the web.
From this screen you can also install Beetailer, which is a service that allows
you to integrate your store into Facebook. This will work only if your server has
a publicly accesible IP address.

The next screen is the Installation Folder, where you can select where Bitnami 
stack will be installed. If the destination directory does not exist, it will 
be created as part of the installation. 

The next screen will prompt you for data necessary to create the initial
admin user: 

Username and password: You will use this information to log-in into the
administrative interface. The password you provide here will also be used to
protect other parts of the installation and to log-in in Magento. Please see 
the section named "Usernames and Passwords" later in this document.

Real name: The name that will be displayed in the application.

Email address: Your email address.

The next screen will vary, depending on whether the ports needed by the
bundled Apache and MySQL are already taken. The default listening port for
Apache is 8080 on Linux or port 80 on WIndows and for MySQL is 3006. If those 
ports are already in use by other applications, you will be prompted for 
alternate ports to use. After this screen, yo will be asked for the machine
hostname or IP address.

Once the information has been entered, the installation will proceed to copy
the files to the target installation directory and configure the different
components. One this process has  been completed, you will see the
'Installation Finished' page. You can choose to launch Bitnami Magento
Stack at this point. If you do so, your default web browser will open and
display the Welcome page for the Bitnami Magento Stack. 

If you received an error message during installation, please refer to
the Troubleshooting section.

The rest of this guide assumes that you installed Bitnami Magento
Stack in /home/user/magento-1.9.1.0-1 on Linux or C:\Program Files\Bitnami Magento Stack
on Windows or /Applications/magento-1.9.1.0-1 on OS X and that you use port 8080 
for Apache on Linux or port 80 on Windows and 3306 for MySQL. 

6. STARTING AND STOPPING BITNAMI MAGENTO STACK

To start/stop/restart application on Linux you can use the included ctlscript.sh
utility, as shown below:

       ./ctlscript.sh (start|stop|restart)
       ./ctlscript.sh (start|stop|restart) mysql
       ./ctlscript.sh (start|stop|restart) apache

  start      - start the service(s)
  stop       - stop  the service(s)
  restart    - restart or start the service(s)

The Bitnami Control Panel is a simple graphical interface included in the OS X stacks that 
can start and stop the Bitnami servers. It is are located in the same installation directory. 
To start the utility, double click the file named ‘control’ from your file browser.

On Windows, you can start and stop Bitnami Magento Stack using the shortcuts created
in the Start Menu, under Programs -> Bitnami -> Bitnami Service 

That will start Apache service. Once started, you can open your
browser and access the following URL on Linux or OS X:

http://127.0.0.1:8080/magento/

And you can access the following URL on Windows:

http://127.0.0.1/magento/

If you selected an alternate port during installation, for example 18080, the
URL will look like:

http://127.0.0.1:18080/magento/

If you are accessing the machine remotely, you will need to replace
127.0.0.1 with the appropriate IP address or hostname.

7. DIRECTORY STRUCTURE

The installation process will create several subfolders under the main
installation directory:

	apache2/: Apache Web server.
	php/: PHP Scripting Language.
	mysql/: MySQL Database.
	apps/
	  magento/: Magento application folder
	    conf/: Magento Apache configuration files
	    htdocs/: Magento PHP application files
	    scripts/: tools to install/uninstall Magento
	    temp/: a temporally folder created after installation (you can delete it safely)
	  phpMyAdmin/: phpMyAdmin application folder (optional)

8. DEFAULT USERNAMES AND PASSWORDS

The Magento administrative user and password are the ones you set at
installation time. 

MySQL admin user is called 'root', and its password is the same as the
Magento administrative user password.

The phpMyAdmin administrative user is 'root'.

For Virtual Appliances and AMIs you can check the default user and password
in the Bitnami download page. The administrator user is 'user' and the
password is 'bitnami1'


9. TROUBLESHOOTING

This version of the Bitnami Magento stack is a preview version, and as
such, may contain a number of bugs and be a little bit rough around the
edges. We are working on the next release, which will contain several
improvements along with expanded documentation. In addition to the resources
provided below, we encourage you to post your questions and suggestions at:

https://community.bitnami.com/

We also encourage you to sign up for our newsletter, which we'll use to
announce new releases and new stacks. To do so, just register at:
https://bitnami.com/newsletter.  

9.1 Installer

# Installer Payload Error 

If you get the following error while trying to run the installer from the
command line:

"Installer payload initialization failed. This is likely due to an
incomplete or corrupt downloaded file" 

The installer binary is not complete, likely because the file was
not downloaded correctly. You will need to download the file and
repeat the installation process. 

9.2 Apache

If you find any problem starting Apache, the first place you should check is
the Apache error log file:

/home/user/magento-1.9.1.0-1/apache2/logs/error.log on Linux or
C:\Program Files\Bitnami Magento Stack\apache2\logs\error.log on Windows or
/Applications/magento-1.9.1.0-1/apache2/logs/error.log.

Most errors are related to not being able to listen to the default port.
Make sure there are no other server programs listening at the same port
before trying to start Apache. Some programs, such as Skype, may also use
port 80. For issues not covered in this Quick Start guide, please refer to
the Apache documentation, which is located at http://httpd.apache.org/docs/  

9.3 MySQL

The following are some common problems: 

# Access denied when trying to connect to MySQL. 

If you get an Access Denied message while trying to connect to
MySQL, make sure you are using the correct username and password. 

# "Can't connect to server" message. 

Make sure that the MySQL daemon is up and running. Remember also that if
during installation you selected a different listening port for MySQL, you
may need to pass that as an extra command line option.

For help in troubleshooting MySQL errors, you may want to reference the
"Problems and Common Errors" section of the MySQL manual, which you will
find at http://dev.mysql.com/doc/


9.4 Magento

# I can't log in as Administrator (or I can't find the administration panel)

The administrative interface of Magento is hidden. You can access it by going
to http://your_domain/magento/index.php/admin

You can also find a link to it in the main page of the stack
(http://your_address/) or in the applications page
(http://your_address/applications.html), if you installed Magento as a module.

For any problem related to Magento, please visit 
http://www.magentocommerce.com/boards/


10. LICENSES

Apache Web Server is distributed under the Apache License v2.0, which
is located at http://www.apache.org/licenses/LICENSE-2.0

MySQL is distributed under the GNU General Public License v2, which is
located at http://www.gnu.org/licenses/old-licenses/gpl-2.0.html

PHP and related libraries are distributed under the PHP License v3.01,
which is located at http://www.php.net/license/3_01.txt

Magento is distributed under the Common Public Attribution License Version 1.0 (CPAL)
which is located at: http://www.opensource.org/licenses/cpal_1.0

curl is distributed under the Curl License, which is located at
http://curl.haxx.se/docs/copyright.html

expat is distributed under the MIT License, which is located at
http://www.jclark.com/xml/copying.txt

gd is distributed under the gd License, which is located at
http://www.boutell.com/gd/manual2.0.33.html

IMAP is distributed under the University of Washington Free-Fork
License, located at http://www.washington.edu/imap/legal.html

jpegsrc is distributed under The Independent JPEG Group's JPEG
software license, which is located at
http://dev.w3.org/cvsweb/Amaya/libjpeg/README?rev=1.2

libiconv is distributed under the Lesser General Public License
(LGPL), located at http://www.gnu.org/copyleft/lesser.html

Freetype is released under The Freetype Project License, that is located at
http://freetype.sourceforge.net/FTL.TXT


changelog.txt
============ CHANGELOG =============

Version 1.9.1.0-1      2015-02-07
* Updated PHP to 5.4.37
* Updated Apache to 2.4.12
* Updated MySQL to 5.5.42 for Linux and Windows
* Updated PageSpeed to 1.9.32.3
* Updated phpMyAdmin to 4.3.7
* Updated OpenSSL to 1.0.1k (Linux and OS X)
* Updated PHP to 5.4.36

Version 1.9.1.0-0      2014-11-28
* Updated Magento to 1.9.1.0
* Updated phpMyAdmin to 4.2.12

Version 1.9.0.1-2      2014-10-21
* (Version only available for OS X)
* Fix MySQL startup issue for OS X Yosemite
* Updated PHP to 5.4.34
* Updated OpenSSL to 1.0.1j
* Added OCI8 module. It requires InstantClient 11.2.
* Updated phpMyAdmin to 4.2.9.1
* Updated MySQL to 5.5.40

Version 1.9.0.1-2      2014-09-10
* Applied patch SUPEE-3762 to fix SOAP API bug
* Updated phpMyAdmin to 4.2.8

Version 1.9.0.1-1      2014-08-27
* Moved application to root URL in Cloud Images and Virtual Machines
* Updated MySQL to 5.5.39
* Updated PHP to 5.4.32
* Updated OpenSSL to 1.0.1i
* Updated phpMyAdmin to 4.2.7.1
* Updated Apache to 2.4.10
* Updated XDebug to 2.2.5
* Updated XCache to 3.1.0
* Updated mongo-php-driver to 1.5.4
* Updated php-couchbase to 1.1.5
* Updated Libxslt to 1.1.28
* Updated Libxml2 to 2.9.1
* Fixed module installation issue

Version 1.9.0.1-0      2014-05-19
* Updated Magento to 1.9.0.1

Version 1.9.0.0-0      2014-05-14
* Updated Magento to 1.9.0.0
* Fixed link to admin panel in welcome page
* Added libv8 and v8js PHP extension for Linux and OS X

Version 1.8.1.0-3      2014-04-24
* Fixed an issue with X-Frame-Options configuration

Version 1.8.1.0-2      2014-04-08
* Updated OpenSSL to 1.0.1g
* Updated phpMyAdmin to 4.1.12
* Improved Apache configuration for PHP-FPM mode
* Updated Apache to 2.4.9

Version 1.8.1.0-1      2014-03-20
* Updated Beetailer to 3.1.0
* Fixed support for REST API
* Updated PHP to 5.4.26
* Updated phpMyAdmin to 4.1.9
* Updated MySQL to 5.5.36
* Updated Varnish to 3.0.5
* Updated phpMyAdmin to 4.1.4
* Updated OpenSSL to 1.0.1f in Linux and OSX

Version 1.8.1.0-0      2013-12-16
* Updated Magento to 1.8.1.0
* Updated phpMyAdmin to 4.1.0
* Updated PHP to 5.4.23
* Updated Magento to 1.8.1.0
* Improved FTP configuration
* Updated phpMyAdmin to 4.0.9

Version 1.8.0.0-2      2013-11-13
* Added Varnish 3.0.3
* Updated PHP to 5.4.21
* Updated phpMyAdmin to 4.0.8

Version 1.8.0.0-1      2013-10-09
* Added PHP-FPM support for cloud images and VMs

Version 1.8.0.0-0      2013-09-26
* Updated Magento to 1.8.0.0
* Updated PHP to 5.4.20
* Configure Magento to use MySQL socket on Linux and OS X

Version 1.7.0.2-4      2013-08-27
* Updated PHP to 5.4.19
* Updated Apache to 2.4.6
* Added logrotate configuration files
* Added imagick PHP extension for Linux and OS X
* Added apr-dbd MySQL driver
* Improved mod_pagespeed settings for better performance (Unix)

Version 1.7.0.2-2      2013-07-16
* Added mod_pagespeed 1.6.29.2 for Linux
* Updated PHP to 5.4.17
* Updated MySQL to 5.5.32

Version 1.7.0.2-1      2013-06-10
* Updated PHP to 5.4.16
* Updated phpMyAdmin to 4.0.3
* Updated Beetailer to 3.0.0
* Updated phpMyAdmin to 4.0.1
* Updated PHP to 5.4.15
* Updated MySQL to 5.5.30
* Updated Apache to 2.4.4
* Updated OpenSSL to 1.0.1e
* Improved the Magento installation by allowing Production
and Development modes.
* Updated Memcache to 2.2.7
* Updated XCache to 3.0.0
* Updated curl to 7.27.0
* Updated AWS SDK to 1.5.13
* Added ModSecurity Apache module 2.6.7
* Added Mail, Net_Socket and Net_SMTP pear packages
* Increased the max allowed size for upload files to 32M

Version 1.7.0.2-0      2012-07-11
* Updated Magento to 1.7.0.2

Version 1.7.0.1-0      2012-06-21
* Updated Magento to 1.7.0.1
* Updated PHP to 5.3.14

Version 1.7.0.0-0      2012-05-02
* Update Magento to 1.7.0.0

Version 1.6.2.0-0      2012-04-27
* Update Magento to 1.6.2.0
* Update PHP to 5.3.9
* Update phpMyAdmin 3.4.9
* Update MySQL to 5.5.16
* Update Apache to 2.2.21
* Update AWS SDK for PHP to 1.5.1

Version 1.6.1.0-0      2011-10-20
* Updated Magento to 1.6.1.0
* Updated phpMyAdmin to 3.4.6

Version 1.6.0.0-2      2011-09-30
* Added mail support with SMTP
* Mark as read notifications
* Fix updateip for Virtual Appliances

Version 1.6.0.0-1      2011-09-08
* Fix issue updating the domain in the application

Version 1.6.0.0-0      2011-08-23
* Update Magento to 1.6.0.0
* Disable MySQL binary logging
* Don't allow to execute the ctlscript.sh script as root 
when the stack is installed as a non root user

Version 1.5.1.0-0      2011-05-16
* Update Magento to 1.5.1.0
* Update phpMyAdmin to 3.3.10
* Update PHP to 5.3.5
* Bundles Beetailer 2.3.5

Version 1.5.0.1-0      2011-04-15
* Initial release, bundles Apache 2.2.17, phpMyAdmin 2.3.3.10,
PHP 5.2.17, MySQL 5.1.56 and Magento 1.5.0.1


php/lib/php/pearcmd.php
<?php
/**
 * PEAR, the PHP Extension and Application Repository
 *
 * Command line interface
 *
 * PHP versions 4 and 5
 *
 * @category   pear
 * @package    PEAR
 * @author     Stig Bakken <ssb@php.net>
 * @author     Tomas V.V.Cox <cox@idecnet.com>
 * @copyright  1997-2009 The Authors
 * @license    http://opensource.org/licenses/bsd-license.php New BSD License
 * @version    CVS: $Id: pearcmd.php 313023 2011-07-06 19:17:11Z dufuz $
 * @link       http://pear.php.net/package/PEAR
 */

ob_end_clean();
if (!defined('PEAR_RUNTYPE')) {
    // this is defined in peclcmd.php as 'pecl'
    define('PEAR_RUNTYPE', 'pear');
}
define('PEAR_IGNORE_BACKTRACE', 1);
/**
 * @nodep Gtk
 */
if ('/home/ubuntu/magento/magento-1.9.1.0-1-64-bit/php/lib/php' != '@'.'include_path'.'@') {
    ini_set('include_path', '/home/ubuntu/magento/magento-1.9.1.0-1-64-bit/php/lib/php');
    $raw = false;
} else {
    // this is a raw, uninstalled pear, either a cvs checkout, or php distro
    $raw = true;
}
@ini_set('allow_url_fopen', true);
if (!ini_get('safe_mode')) {
    @set_time_limit(0);
}
ob_implicit_flush(true);
@ini_set('track_errors', true);
@ini_set('html_errors', false);
@ini_set('magic_quotes_runtime', false);
$_PEAR_PHPDIR = '#$%^&*';
set_error_handler('error_handler');

$pear_package_version = "1.9.4";

require_once 'PEAR.php';
require_once 'PEAR/Frontend.php';
require_once 'PEAR/Config.php';
require_once 'PEAR/Command.php';
require_once 'Console/Getopt.php';


PEAR_Command::setFrontendType('CLI');
$all_commands = PEAR_Command::getCommands();

// remove this next part when we stop supporting that crap-ass PHP 4.2
if (!isset($_SERVER['argv']) && !isset($argv) && !isset($HTTP_SERVER_VARS['argv'])) {
    echo 'ERROR: either use the CLI php executable, or set register_argc_argv=On in php.ini';
    exit(1);
}

$argv = Console_Getopt::readPHPArgv();
// fix CGI sapi oddity - the -- in pear.bat/pear is not removed
if (php_sapi_name() != 'cli' && isset($argv[1]) && $argv[1] == '--') {
    unset($argv[1]);
    $argv = array_values($argv);
}
$progname = PEAR_RUNTYPE;
array_shift($argv);
$options = Console_Getopt::getopt2($argv, "c:C:d:D:Gh?sSqu:vV");
if (PEAR::isError($options)) {
    usage($options);
}

$opts = $options[0];

$fetype = 'CLI';
if ($progname == 'gpear' || $progname == 'pear-gtk') {
    $fetype = 'Gtk';
} else {
    foreach ($opts as $opt) {
        if ($opt[0] == 'G') {
            $fetype = 'Gtk';
        }
    }
}
//Check if Gtk and PHP >= 5.1.0
if ($fetype == 'Gtk' && version_compare(phpversion(), '5.1.0', '>=')) {
    $fetype = 'Gtk2';
}

$pear_user_config = '';
$pear_system_config = '';
$store_user_config = false;
$store_system_config = false;
$verbose = 1;

foreach ($opts as $opt) {
    switch ($opt[0]) {
        case 'c':
            $pear_user_config = $opt[1];
            break;
        case 'C':
            $pear_system_config = $opt[1];
            break;
    }
}

PEAR_Command::setFrontendType($fetype);
$ui = &PEAR_Command::getFrontendObject();
$config = &PEAR_Config::singleton($pear_user_config, $pear_system_config);

if (PEAR::isError($config)) {
    $_file = '';
    if ($pear_user_config !== false) {
       $_file .= $pear_user_config;
    }
    if ($pear_system_config !== false) {
       $_file .= '/' . $pear_system_config;
    }
    if ($_file == '/') {
        $_file = 'The default config file';
    }
    $config->getMessage();
    $ui->outputData("ERROR: $_file is not a valid config file or is corrupted.");
    // We stop, we have no idea where we are :)
    exit(1);
}

// this is used in the error handler to retrieve a relative path
$_PEAR_PHPDIR = $config->get('php_dir');
$ui->setConfig($config);
PEAR::setErrorHandling(PEAR_ERROR_CALLBACK, array($ui, "displayFatalError"));
if (ini_get('safe_mode')) {
    $ui->outputData('WARNING: running in safe mode requires that all files created ' .
        'be the same uid as the current script.  PHP reports this script is uid: ' .
        @getmyuid() . ', and current user is: ' . @get_current_user());
}

$verbose = $config->get("verbose");
$cmdopts = array();

if ($raw) {
    if (!$config->isDefinedLayer('user') && !$config->isDefinedLayer('system')) {
        $found = false;
        foreach ($opts as $opt) {
            if ($opt[0] == 'd' || $opt[0] == 'D') {
                $found = true; // the user knows what they are doing, and are setting config values
            }
        }
        if (!$found) {
            // no prior runs, try to install PEAR
            if (strpos(dirname(__FILE__), 'scripts')) {
                $packagexml = dirname(dirname(__FILE__)) . DIRECTORY_SEPARATOR . 'package2.xml';
                $pearbase = dirname(dirname(__FILE__));
            } else {
                $packagexml = dirname(__FILE__) . DIRECTORY_SEPARATOR . 'package2.xml';
                $pearbase = dirname(__FILE__);
            }
            if (file_exists($packagexml)) {
                $options[1] = array(
                    'install',
                    $packagexml
                );
                $config->set('php_dir', $pearbase . DIRECTORY_SEPARATOR . 'php');
                $config->set('data_dir', $pearbase . DIRECTORY_SEPARATOR . 'data');
                $config->set('doc_dir', $pearbase . DIRECTORY_SEPARATOR . 'docs');
                $config->set('test_dir', $pearbase . DIRECTORY_SEPARATOR . 'tests');
                $config->set('ext_dir', $pearbase . DIRECTORY_SEPARATOR . 'extensions');
                $config->set('bin_dir', $pearbase);
                $config->mergeConfigFile($pearbase . 'pear.ini', false);
                $config->store();
                $config->set('auto_discover', 1);
            }
        }
    }
}
foreach ($opts as $opt) {
    $param = !empty($opt[1]) ? $opt[1] : true;
    switch ($opt[0]) {
        case 'd':
            if ($param === true) {
                die('Invalid usage of "-d" option, expected -d config_value=value, ' .
                    'received "-d"' . "\n");
            }
            $possible = explode('=', $param);
            if (count($possible) != 2) {
                die('Invalid usage of "-d" option, expected -d config_value=value, received "' .
                    $param . '"' . "\n");
            }
            list($key, $value) = explode('=', $param);
            $config->set($key, $value, 'user');
            break;
        case 'D':
            if ($param === true) {
                die('Invalid usage of "-d" option, expected -d config_value=value, ' .
                    'received "-d"' . "\n");
            }
            $possible = explode('=', $param);
            if (count($possible) != 2) {
                die('Invalid usage of "-d" option, expected -d config_value=value, received "' .
                    $param . '"' . "\n");
            }
            list($key, $value) = explode('=', $param);
            $config->set($key, $value, 'system');
            break;
        case 's':
            $store_user_config = true;
            break;
        case 'S':
            $store_system_config = true;
            break;
        case 'u':
            $config->remove($param, 'user');
            break;
        case 'v':
            $config->set('verbose', $config->get('verbose') + 1);
            break;
        case 'q':
            $config->set('verbose', $config->get('verbose') - 1);
            break;
        case 'V':
            usage(null, 'version');
        case 'c':
        case 'C':
            break;
        default:
            // all non pear params goes to the command
            $cmdopts[$opt[0]] = $param;
            break;
    }
}

if ($store_system_config) {
    $config->store('system');
}

if ($store_user_config) {
    $config->store('user');
}

$command = (isset($options[1][0])) ? $options[1][0] : null;
if (empty($command) && ($store_user_config || $store_system_config)) {
    exit;
}

if ($fetype == 'Gtk' || $fetype == 'Gtk2') {
    if (!$config->validConfiguration()) {
        PEAR::raiseError('CRITICAL ERROR: no existing valid configuration files found in files ' .
            "'$pear_user_config' or '$pear_system_config', please copy an existing configuration" .
            'file to one of these locations, or use the -c and -s options to create one');
    }
    Gtk::main();
} else do {
    if ($command == 'help') {
        usage(null, @$options[1][1]);
    }

    if (!$config->validConfiguration()) {
        PEAR::raiseError('CRITICAL ERROR: no existing valid configuration files found in files ' .
            "'$pear_user_config' or '$pear_system_config', please copy an existing configuration" .
            'file to one of these locations, or use the -c and -s options to create one');
    }

    PEAR::pushErrorHandling(PEAR_ERROR_RETURN);
    $cmd = PEAR_Command::factory($command, $config);
    PEAR::popErrorHandling();
    if (PEAR::isError($cmd)) {
        usage(null, @$options[1][0]);
    }

    $short_args = $long_args = null;
    PEAR_Command::getGetoptArgs($command, $short_args, $long_args);
    array_shift($options[1]);
    $tmp = Console_Getopt::getopt2($options[1], $short_args, $long_args);

    if (PEAR::isError($tmp)) {
        break;
    }

    list($tmpopt, $params) = $tmp;
    $opts = array();
    foreach ($tmpopt as $foo => $tmp2) {
        list($opt, $value) = $tmp2;
        if ($value === null) {
            $value = true; // options without args
        }

        if (strlen($opt) == 1) {
            $cmdoptions = $cmd->getOptions($command);
            foreach ($cmdoptions as $o => $d) {
                if (isset($d['shortopt']) && $d['shortopt'] == $opt) {
                    $opts[$o] = $value;
                }
            }
        } else {
            if (substr($opt, 0, 2) == '--') {
                $opts[substr($opt, 2)] = $value;
            }
        }
    }

    $ok = $cmd->run($command, $opts, $params);
    if ($ok === false) {
        PEAR::raiseError("unknown command `$command'");
    }

    if (PEAR::isError($ok)) {
        PEAR::setErrorHandling(PEAR_ERROR_CALLBACK, array($ui, "displayFatalError"));
        PEAR::raiseError($ok);
    }
} while (false);

// {{{ usage()

function usage($error = null, $helpsubject = null)
{
    global $progname, $all_commands;
    $stdout = fopen('php://stdout', 'w');
    if (PEAR::isError($error)) {
        fputs($stdout, $error->getMessage() . "\n");
    } elseif ($error !== null) {
        fputs($stdout, "$error\n");
    }

    if ($helpsubject != null) {
        $put = cmdHelp($helpsubject);
    } else {
        $put = "Commands:\n";
        $maxlen = max(array_map("strlen", $all_commands));
        $formatstr = "%-{$maxlen}s  %s\n";
        ksort($all_commands);
        foreach ($all_commands as $cmd => $class) {
            $put .= sprintf($formatstr, $cmd, PEAR_Command::getDescription($cmd));
        }
        $put .=
            "Usage: $progname [options] command [command-options] <parameters>\n".
            "Type \"$progname help options\" to list all options.\n".
            "Type \"$progname help shortcuts\" to list all command shortcuts.\n".
            "Type \"$progname help <command>\" to get the help for the specified command.";
    }
    fputs($stdout, "$put\n");
    fclose($stdout);

    if ($error === null) {
        exit(0);
    }
    exit(1);
}

function cmdHelp($command)
{
    global $progname, $all_commands, $config;
    if ($command == "options") {
        return
        "Options:\n".
        "     -v         increase verbosity level (default 1)\n".
        "     -q         be quiet, decrease verbosity level\n".
        "     -c file    find user configuration in `file'\n".
        "     -C file    find system configuration in `file'\n".
        "     -d foo=bar set user config variable `foo' to `bar'\n".
        "     -D foo=bar set system config variable `foo' to `bar'\n".
        "     -G         start in graphical (Gtk) mode\n".
        "     -s         store user configuration\n".
        "     -S         store system configuration\n".
        "     -u foo     unset `foo' in the user configuration\n".
        "     -h, -?     display help/usage (this message)\n".
        "     -V         version information\n";
    } elseif ($command == "shortcuts") {
        $sc = PEAR_Command::getShortcuts();
        $ret = "Shortcuts:\n";
        foreach ($sc as $s => $c) {
            $ret .= sprintf("     %-8s %s\n", $s, $c);
        }
        return $ret;

    } elseif ($command == "version") {
        return "PEAR Version: ".$GLOBALS['pear_package_version'].
               "\nPHP Version: ".phpversion().
               "\nZend Engine Version: ".zend_version().
               "\nRunning on: ".php_uname();

    } elseif ($help = PEAR_Command::getHelp($command)) {
        if (is_string($help)) {
            return "$progname $command [options] $help\n";
        }

        if ($help[1] === null) {
            return "$progname $command $help[0]";
        }

        return "$progname $command [options] $help[0]\n$help[1]";
    }

    return "Command '$command' is not valid, try '$progname help'";
}

// }}}

function error_handler($errno, $errmsg, $file, $line, $vars) {
    if ((defined('E_STRICT') && $errno & E_STRICT) || (defined('E_DEPRECATED') &&
          $errno & E_DEPRECATED) || !error_reporting()) {
        if (defined('E_STRICT') && $errno & E_STRICT) {
            return; // E_STRICT
        }
        if (defined('E_DEPRECATED') && $errno & E_DEPRECATED) {
            return; // E_DEPRECATED
        }
        if ($GLOBALS['config']->get('verbose') < 4) {
            return false; // @silenced error, show all if debug is high enough
        }
    }
    $errortype = array (
        E_ERROR   =>  "Error",
        E_WARNING   =>  "Warning",
        E_PARSE   =>  "Parsing Error",
        E_NOTICE   =>  "Notice",
        E_CORE_ERROR  =>  "Core Error",
        E_CORE_WARNING  =>  "Core Warning",
        E_COMPILE_ERROR  =>  "Compile Error",
        E_COMPILE_WARNING =>  "Compile Warning",
        E_USER_ERROR =>  "User Error",
        E_USER_WARNING =>  "User Warning",
        E_USER_NOTICE =>  "User Notice"
    );
    $prefix = $errortype[$errno];
    global $_PEAR_PHPDIR;
    if (stristr($file, $_PEAR_PHPDIR)) {
        $file = substr($file, strlen($_PEAR_PHPDIR) + 1);
    } else {
        $file = basename($file);
    }
    print "\n$prefix: $errmsg in $file on line $line\n";
    return false;
}


/*
 * Local variables:
 * tab-width: 4
 * c-basic-offset: 4
 * indent-tabs-mode: nil
 * mode: php
 * End:
 */
// vim600:syn=php